dataproc/dataproc.py

```diff
@@ -0,0 +1,167 @@
+
+
+# The channels this script will import.
+CHANNELS = ["nightly", "beta", "release"]
+# The number of historical versions to import for each channel.
+NUM_VERSIONS = 3
+
+
+PRODUCT_DETAILS_URL = "https://product-details.mozilla.org/1.0/firefox_versions.json"
+SOURCE_BIGQUERY_TABLE = "moz-fx-data-derived-datasets.telemetry.client_probe_counts_v2"
+
+
+def retry(exceptions, tries=3, delay=1, backoff=2):
+    def fn(f):
+        @wraps(f)
+        def fn_retry(*args, **kwargs):
+            mtries, mdelay = tries, delay
+            while mtries > 1:
+                try:
+                    return f(*args, **kwargs)
+                except exceptions as e:
+                    logger.error(
+                        "Error: {e}, Retrying in {mdelay} seconds...".format(
+                            e=e, mdelay=mdelay
+                        )
+                    )
+                    time.sleep(mdelay)
+                    mtries -= 1
+                    mdelay *= backoff
+            else:
+                logger.warning("All retries have been consumed.")
+
+            return f(*args, **kwargs)
+
+        return fn_retry  # true decorator
+
+    return fn
+
+
+@retry((DeadlineExceeded, ServiceUnavailable), tries=8, delay=1, backoff=2)
+def commit_batch(batch):
+    batch.commit()
+
+
+def batch_insert(group):
+    firestore_client = firestore.Client()
+
+    logger.info("Processing batch {}.".format(group[0]))
+
+    batch = firestore_client.batch()
+    for row in group[1]:
+        # Create the document we will store from the columns selected.
+        doc = row.asDict()
+        # Remap the aggregates instead of ``Row`` objects.
+        doc["aggregates"] = {
+            d["key"]: round(d["value"], 4)
+            for d in map(lambda x: x.asDict(), doc["aggregates"])
+        }
+        # Pop the `channel` and `app_version` for use in the collection.
+        channel = doc.pop("channel")
+        app_version = doc.pop("app_version")
+        # Generate a document ID so this can be updated each run.
+        doc_id = hashlib.blake2b(
+            "{metric}-{build_id}-{os}-{agg_type}-{key}".format(**doc).encode()
+        ).hexdigest()
+        # Store this document in the collection by channel+version.
+        collection = "{channel}-{app_version}".format(
+            channel=channel, app_version=app_version
+        )
+        batch.set(firestore_client.collection(collection).document(doc_id), doc)
+
+    commit_batch(batch)
+
+    logger.info("Processed batch {}.".format(group[0]))
+
+
+def extract(query):
+    logger.info(query)
+
+    bigquery_client = bigquery.Client()
+
+    # Execute query to get destination dataset.
+    job = bigquery_client.query(query)
+    job.result()  # Wait for job to complete.
+
+    logger.info("Query complete, processing via Spark.")
+
+    (
+        SparkSession.builder.appName("export_to_firestore")
+        .getOrCreate()
+        .read.format("bigquery")
+        .option("parallelism", 6)
```

Comment:
this should probably be set to 0, which lets the api decide based on volume, instead of a static number. if that doesn't work, try it at a variety of numbers, like 20, 50, and 100, and pick the highest number that reliably completes. I would expect at least 20 to work reliably. I found that setting it too high resulted in Out of Memory exceptions for main summary, but that job had complications that this might not.

Reply:
alternately, make this config option, so that it doesn't block merging the PR.